Clarify the semantics of `Map.equals` and `Set.equals` in Scaladoc

lrytz opened the pull request:
Also change the overrides of `SortedMap.equals` and `SortedSet.equals`
to check for key equivalence according to the ordering, instead of
key equality.

Fixes https://github.com/scala/bug/issues/12228

lrytz committed:
Clarify the semantics of `Map.equals` and `Set.equals` in Scaladoc

Also change the overrides of `SortedMap.equals` and `SortedSet.equals`
to check for key equivalence according to the ordering, instead of
key equality.

NthPortal commented on src/library/scala/collection/SortedSet.scala:
```diff
@@ -29,14 +29,14 @@ trait SortedSet[A] extends Set[A]
   override def equals(that: Any): Boolean = that match {
     case _ if this eq that.asInstanceOf[AnyRef] => true
-    case ss: SortedSet[_] if ss.ordering == this.ordering =>
+    case ss: SortedSet[A] if ss.ordering == this.ordering =>
```
I'm surprised this doesn't need an `@unchecked` annotation

lrytz replied:
Me too...

```
scala> def f(a: Any) = a match { case s: Set[Int] => 0 }
                                         ^
       warning: non-variable type argument Int in type pattern scala.collection.immutable.Set[Int] (the underlying of Set[Int]) is unchecked since it is eliminated by erasure
def f(a: Any): Int

scala> class C[A] { def f(a: Any) = a match { case s: Set[A] => 0 } }
class C
```

NthPortal replied:
is it possible this is a bug/regression? I would swear it used to warn...

lrytz replied:
I'm not sure, but it's worth figuring out. It's used in many places across the collections...

martijnhoekstra replied:
This does actually look unsound to me. I may have time to construct a use that leads to a ClassCastException later. I'm not sure what we can do about it.

lrytz replied:
I also have doubts, but couldn't come up with a counterexample...

NthPortal replied:
as long as `ss.ordering` has an `equals` method consistent with the contract of equality, I think it should be sound

martijnhoekstra replied:
If you allow `equals` to be inconsistent with the contract, it seems you can definitely trigger a ClassCastException. That's not good, but maybe worth it to fix the issue per so don't do that. Otherwise, I'm not too sure, my mental code-generation and bytecode checking isn't that reliable.

smarter replied:
fwiw, when the standard library is compiled with dotty it generates a bunch of unchecked warnings which I don't think anyone has inspected, in this PR this includes the line in question: https://travis-ci.com/github/scala/scala/jobs/497265334

SethTisue replied:
I'm hitting "Resolve" because I think this is resolved enough to merge this PR, even if the bigger picture remains unclear.

dwijnand replied:
Yeah, definitely the pattern typer letting us down here...

```scala
scala> class C[A](ord: Ordering[A], x: A) { def f(a: Any) = a match { case s: Set[A] => ord.equiv(s.iterator.next(), x) case _ => false } }
class C

scala> new C[String](implicitly[Ordering[String]], "foo").f(Set(1))
java.lang.ClassCastException: java.lang.Integer cannot be cast to java.lang.String
  at scala.math.Ordering$String$.compare(Ordering.scala:576)
  at scala.math.Ordering.equiv(Ordering.scala:103)
  at scala.math.Ordering.equiv$(Ordering.scala:103)
  at scala.math.Ordering$String$.equiv(Ordering.scala:576)
  at C.f(<console>:1)
  ... 32 elided
```

martijnhoekstra commented on src/library/scala/collection/SortedSet.scala:
```diff
@@ -29,14 +29,14 @@ trait SortedSet[A] extends Set[A]
         var allEqual = true
         while (allEqual && i1.hasNext)
-          allEqual = i1.next() == i2.next()
+          allEqual = ordering.equiv(i1.next(), i2.next())
```
Is this sound when `that.A` is different from `A` but somehow `ss.ordering == this.ordering`?

martijnhoekstra reviewed (commented):
~

martijnhoekstra commented on src/library/scala/collection/Set.scala:
```diff
@@ -28,6 +28,37 @@ trait Set[A]
+   * Equality of sets is implemented using the lookup method [[contains]]. This method returns `true` if
+   *   - the argument `o` is a `Set`,
+   *   - the two sets have the same [[size]], and
+   *   - for every `key` this set, `other.contains(key) == true`.
```
Maybe `key` should be `element` for `Set` (then `k1`, `k2` etc. should be `e1`, `e2`, etc.)

martijnhoekstra commented on src/library/scala/collection/Set.scala:
```diff
@@ -28,6 +28,37 @@ trait Set[A]
+   * }}}
+   *
+   *
+   * @param o The set to which this set is compared
```
```suggestion
   * @param that The set to which this set is compared
```

Ichoran commented:
I don't understand why `collection.Map.equals` isn't broken.  `getOrElse` is generic in the key value; the key is not typed as `AnyRef`.  I agree that the `SortedMap` one is okay, but `collection.Map` has the same cast, and I can't see why it's safe except by accident of the existing implementations.  (E.g. `HashMap` uses only `hashCode` and `==`, which are defined on everything.)

lrytz commented:
You're absolutely right... 🤯

```
scala> class M extends Map[String, String] {
     |   def iterator = Iterator("" -> "")
     |   def removed(key: String) = this
     |   def updated[V1 >: String](key: String, value: V1) = this
     |   def get(key: String) = if (key.trim == key) Some("") else None
     | }
class M

scala> Map(1 -> 2) == new M
java.lang.ClassCastException: java.lang.Integer cannot be cast to java.lang.String
```

lrytz commented:
Or even more built-in:

```
scala> Set("hai") == scala.collection.immutable.BitSet(1)
java.lang.ClassCastException: java.lang.String cannot be cast to java.lang.Integer
```

Ichoran commented:
Oh, the bitset one is a good example.  But this is disturbing.  Do we have any way to fix this without either trashing performance or binary compatibility?

Note that we can't even just catch the exception and call it false, because

```
collection.immutable.BitSet(1) == Set(1L)
```

is true, but if you swap the order you get a `ClassCastException`.

Ichoran commented:
I can think of three classes of fix, none of which are binary compatible, and all of which have space or time drawbacks.

1. Add a typeclass that witnesses the key type.  Then you can match on it.  Downside: requires additional storage for every set and map; numbers of different types might be hard to handle.
2. Add a `containsAny(a: Any)` method to sets (also `getAny`, `getAnyOrElse` on maps) that you call instead of `contains` (or `getOrElse`).  Downside: complicates the interface for no good reason; additional class check on every lookup; numbers of different types are incompatible instead of getting special treatment.  Mitigation: the JVM maybe has to do the check anyway or you wouldn't get a CCE, so maybe there's no performance penalty?
3. Cooperative checking of the collection type to avoid calling that is in practice (as opposed to just potentially) specialized by key type.  Downside: can't be cooperative with user collections; not clear that this method can be used if both collections have specialized keys (because you couldn't necessarily tell that they were compatible e.g. if they're both ordered).

If anyone can think of anything better, that would be lovely.

Note: this fix doesn't make sense in this ticket, but if we don't at least know what the fix will be (or if we will decline a fix), I don't think we can clarify the semantics: we might have to change them depending on what we decide to do here.

RubenVerg reviewed (approved):
looks pretty clear to me

ps can anyone else approve now or did i do it for everybody in the team?

lrytz commented:
Now I realize why there used to be a `catch { case _: ClassCastException => false }` in `Map.equals`... I removed that not long ago: https://github.com/scala/scala/pull/9117/files.

@Ichoran what do you think about bringing this back..?

viktorklang commented on src/library/scala/collection/Map.scala:
```diff
@@ -29,6 +29,37 @@ trait Map[K, +V]
+   */
   override def equals(o: Any): Boolean =
     (this eq o.asInstanceOf[AnyRef]) || (o match {
       case map: Map[K, _] if map.canEqual(this) =>
```
@lrytz Do any of the Map implementations even implement `canEqual`, because I think the this.size == that.size would make sense to put there.

lrytz replied:
What would be the advantage of doing that?

As far as I understand, the goal of `canEqual` is to allow subclasses to define what they are equal to, so that equality remains symmetric. https://www.artima.com/pins1ed/object-equality.html

viktorklang replied:
@lrytz Can a map ever equal another map whose size is dissimilar?

lrytz replied:
No, the `equals` implementation compares the sizes first. What would be the advantage of moving that check into `canEquals`?

viktorklang replied:
@lrytz I guess my point swas that the canEqual check doesn't really do much here, and you could either move the size-logic into canEqual or perhaps move it into the guard.

Ichoran commented:
> Now I realize why there used to be a `catch { case _: ClassCastException => false }` in `Map.equals`... I removed that not long ago: https://github.com/scala/scala/pull/9117/files.
> 
> @Ichoran what do you think about bringing this back..?

@lrytz - I think it's better than not bringing it back, but it doesn't fully solve the problem.  (It's also really slow when it happens, but I guess we can let people figure that out with profiling.)

smarter commented:
ClassCastException is undefined behavior on scala.js by default, so not depending on it would be wise: https://www.scala-js.org/doc/semantics.html

lrytz committed:
Catch ClassCastException in Map.equals / Set.equals

`Map.equals` calls `otherMap.getOrElse(keyFromThisMap, ...)` which
fails with a `ClassCastException` if `otherMap` assumes a certain
key type. Similarly for `Set.equals`.

This is not a good solution, but nothing better is known as of now
given the binary compatibility constraints, see discussion in PR
9565.

lrytz commented:
Pushed the two `catch { case _: ClassCastException => false }`

SethTisue commented:
Reading _War and Peace_ wouldn't take long enough, so instead I re-read all the discussion on the original ticket and here. 

In short, I think it's pretty clear should merge this PR. It remains an open question whether a different design might be better in the long run, but this PR strictly improves on the status quo, and nobody has made an alternate proposal that is both clearly better and clearly actually doable in a 2.13.x release.

So yeah, let's merge, but that doesn't mean we're slamming the door shut forever about what else might happen in the future.
